qadence/mitigations/analog_zne.py

```diff
@@ -7,33 +7,133 @@
 
 from qadence import BackendName
-from qadence.backend import ConvertedCircuit, ConvertedObservable
 from qadence.backends.api import backend_factory
 from qadence.backends.pulser.backend import Backend
 from qadence.blocks import block_to_tensor
+from qadence.blocks.abstract import AbstractBlock
+from qadence.blocks.analog import ConstantAnalogRotation, WaitBlock
+from qadence.circuit import QuantumCircuit
 from qadence.measurements import Measurements
 from qadence.mitigations import Mitigations
 from qadence.noise import Noise
+from qadence.operations import AnalogRot
+from qadence.transpile import apply_fn_to_blocks
 from qadence.utils import Endianness
 
 
-def zne(noise: Noise, zne_datasets: list) -> list:
+def zne_pulse(
+    stretches: Tensor, zne_datasets: list[list], n_observables: int, n_params: int
+) -> Tensor:
+    poly_fits = []
+    for o in range(n_observables):
+        batched_observables: list = []
+        for p in range(1 if n_params == 0 else n_params):
+            rearranged_dataset = [s[o][p] for s in zne_datasets]
+            # Polynomial fit function.
+            poly_fit = np.poly1d(np.polyfit(stretches, rearranged_dataset, len(stretches) - 1))
+            # Return the zero-noise extrapolated value.
+            batched_observables.append(poly_fit(0.0))
+        poly_fits.append(batched_observables)
+
+    return torch.tensor(poly_fits)
+
+
+def zne_noise(noise_probas: Tensor, zne_datasets: list[list]) -> Tensor:
+    from matplotlib import pyplot as plt
+
     # Rearrange the dataset by selecting each element in the batches.
-    noise_probas = noise.options.get("noise_probas")
     poly_fits = []
     for datasets in zne_datasets:  # Loop over batches of observables.
+        batched_fits = []
         for i in range(len(datasets[0])):  # Loop over batch length.
             rearranged_dataset = [s[i] for s in datasets]
+            plt.plot(noise_probas, rearranged_dataset, "o")
             # Polynomial fit function.
-            poly_fits.append(np.poly1d(np.polyfit(noise_probas, rearranged_dataset, 4)))
+            poly_fit = np.poly1d(
+                np.polyfit(noise_probas, rearranged_dataset, len(noise_probas) - 1)
+            )
+            # Return the zero-noise extrapolated value.
+            plt.plot(noise_probas, poly_fit(noise_probas), "-")
+            plt.plot(0.0, poly_fit(0.0), "*")
+            plt.show()
+            batched_fits.append(poly_fit(0.0))
+        poly_fits.append(batched_fits)
+
+    return torch.tensor(poly_fits)
 
-    return list(map(lambda p: p(0.0), poly_fits))  # Return the zero-noise fitted value.
+
+def pulse_experiment(
+    backend: Backend,
+    circuit: QuantumCircuit,
+    observables: list[AbstractBlock],
+    param_values: dict[str, Tensor],
+    noise: Noise,
+    stretches: dict[str, Tensor],
+    endianness: Endianness,
+    state: Tensor | None = None,
+) -> Tensor:
+    def mutate_params(block: AbstractBlock, stretch: float) -> AbstractBlock:
```

Comment:
OK, but here the main concern is analog ZNE. I've added your comment to #142 so that it is under our radar. Feel free to comment/detail more there if you wish.